### PR TITLE
Fixes oxyloss icon showing up when someone has the self respiration symptom

### LINKED ISSUE #42237: Self-respiration symptom doesn't prevent oxyloss icon
**Terry**

**Round ID:** 
99785 

**Testmerges:**
#41994 #42155

**Reproduction:**
1. Get a self respiration virus. 
2. Wait for it to reach stage 4. 
3. Go to lavaland without internals. 

The blue O2-icon appears and starts flashing. You properly don't take any oxyloss damage, but the icon makes people think the self respiration is not working, so they put their internals back on. 

https://github.com/tgstation/tgstation/blob/master/code/datums/diseases/advance/symptoms/oxygen.dm

Don't know if intended or not, but I feel it's an issue regardless.

### PR DESCRIPTION
**Pull request by granpawalton**
:cl: GranpaWalton
fix: Oxyloss icon no-longer shows up when someone has the self respiration symptom
tweak: The self respiration now makes you not contract diseases through the air and not breathe in smoke
/:cl:

Fixes #42237 , im still pretty new to this code base so please tell me if I could have done something in a better way

### CODE REVIEW

File: code/datums/diseases/advance/symptoms/oxygen.dm [hunk 1]
```diff
@@ -28,13 +28,16 @@ Bonus
...
+	var/already_NOBREATH = FALSE
 	threshold_desc = "<b>Resistance 8:</b>Additionally regenerates lost blood.<br>"
 
 /datum/symptom/oxygen/Start(datum/disease/advance/A)
 	if(!..())
 		return
 	if(A.properties["resistance"] >= 8) //blood regeneration
 		regenerate_blood = TRUE
+	if(A.affected_mob.has_trait(TRAIT_NOBREATH)) // with inorganic biology one could already have NOBREATH
+		already_NOBREATH = TRUE
```
Comment (coiax) [inline code/datums/diseases/advance/symptoms/oxygen.dm]: You're using the trait system wrong here. One of the strengths of traits as opposed to flags is that they track SOURCES of a trait. So you don't need to keep track of the mob already having the trait, you've just to got add (and remove) your trait with its own unique source.

Something like:
// When symptom means you don't need to breath
mob.add_trait(TRAIT_NOBREATH, DISEASE_TRAIT)

// When the disease is cured or the stage goes down
mob.remove_trait(TRAIT_NOBREATH, DISEASE_TRAIT)

You'll need to define DISEASE_TRAIT in code/__DEFINES/traits.dm, near the other sources of traits.

Comment (granpawalton) [reply code/datums/diseases/advance/symptoms/oxygen.dm]: It should use the trait system correctly now

File: code/datums/diseases/_disease.dm [hunk 2]
```diff
@@ -65,15 +65,17 @@
...
 	else
 		if(prob(cure_chance))
-			stage = max(stage - 1, 1)
+			update_stage(max(stage - 1, 1), src)
 
 	if(disease_flags & CURABLE)
 		if(cure && prob(cure_chance))
 			cure()
 
+/datum/disease/proc/update_stage(new_stage, datum/disease/A)
```
Comment (4dplanner) [inline code/datums/diseases/_disease.dm]: A is just src, unnecessary

Comment (granpawalton) [reply code/datums/diseases/_disease.dm]: Thanks it should no longer pass it as a parameter

File: code/datums/diseases/advance/symptoms/symptoms.dm [hunk 3]
```diff
@@ -60,6 +60,11 @@
 		next_activation = world.time + rand(symptom_delay_min * 10, symptom_delay_max * 10)
 		return TRUE
 
+/datum/symptom/proc/change_stage(new_stage, datum/disease/advance/A)
```
[the remark this answers is not in the document]
Comment (granpawalton) [reply code/datums/diseases/advance/symptoms/symptoms.dm]: ill do that, i was kinda lost on how to name it tbh

File: code/datums/diseases/advance/advance.dm [hunk 4]
```diff
@@ -89,6 +89,12 @@
 		for(var/datum/symptom/S in symptoms)
 			S.Activate(src)
 
+// Tell symptoms stage changed
+/datum/disease/advance/update_stage(new_stage)
+	stage = new_stage
```
Comment (Cyberboss) [inline code/datums/diseases/advance/advance.dm]:
```suggestion
	..()
```

File: code/datums/diseases/advance/symptoms/symptoms.dm [hunk 5]
```diff
@@ -60,6 +60,11 @@
 		next_activation = world.time + rand(symptom_delay_min * 10, symptom_delay_max * 10)
 		return TRUE
 
+/datum/symptom/proc/on_stage_change(new_stage, datum/disease/advance/A)
+	if(neutered)
+		return FALSE
+	return TRUE
```
Comment (Cyberboss) [inline code/datums/diseases/advance/symptoms/symptoms.dm]: This return value isn't used?

File: code/datums/diseases/advance/symptoms/oxygen.dm [hunk 6]
```diff
@@ -50,3 +50,18 @@ Bonus
...
+	switch(A.stage)
+		if(3)
+			M.remove_trait(TRAIT_NOBREATH, DISEASE_TRAIT)
+		if(4)
+			M.add_trait(TRAIT_NOBREATH, DISEASE_TRAIT)
+
+/datum/symptom/oxygen/End(datum/disease/advance/A)
+	if(!..())
+		return
+	A.affected_mob.remove_trait(TRAIT_NOBREATH, DISEASE_TRAIT)
```
Comment (Cyberboss) [inline code/datums/diseases/advance/symptoms/oxygen.dm]: Should only do this if stage >= 4

**Review by Cyberboss — CHANGES_REQUESTED** [review]

**Comment by granpawalton** [comment]
oh i didnt know NOBREATH did that ill update it

**Comment by 4dplanner** [comment]
Don't just take my word for it, I'm away from PC so can't check right now.

File: code/datums/diseases/advance/symptoms/symptoms.dm (hunk 5 above)
Comment (granpawalton) [reply code/datums/diseases/advance/symptoms/symptoms.dm]: isnt it used on oxygen.dm line 55 to tell it not to add the trait if it is neutered?

**Comment by granpawalton** [comment]
you were right thanks for the heads up

File: code/datums/diseases/advance/symptoms/symptoms.dm (hunk 5 above)
Comment (ExcessiveUseOfCobblestone) [reply code/datums/diseases/advance/symptoms/symptoms.dm]: If you're giving the base proc return values you should do so for children overwrites.

Comment (granpawalton) [reply code/datums/diseases/advance/symptoms/symptoms.dm]: im confused, the code that is handled similair in symptoms use the base proc as a means of checking if it is neutered without having return values on their children shouldn't i implement it the same way or would giving the children return values improve the code in a way i am not seeing?

Comment (ExcessiveUseOfCobblestone) [reply code/datums/diseases/advance/symptoms/symptoms.dm]: It's bikeshed, I just like return values

Comment (ExcessiveUseOfCobblestone) [reply code/datums/diseases/advance/symptoms/symptoms.dm]: although i'm assuming it caused the issue here.

Comment (granpawalton) [reply code/datums/diseases/advance/symptoms/symptoms.dm]: ill add them then i just didnt know if it was like a dm thing that made it matter

**Review by Cyberboss — APPROVED** [review]

**Pull request merged**